core/domain/user_jobs_one_off.py

```diff
@@ -468,6 +468,184 @@ def reduce(key, values):
+        models_linked_with_user_settings_model = [
+            ('CompletedActivitiesModel', completed_activities_model),
+            ('IncompleteActivitiesModel', incomplete_activities_model),
+            ('LearnerPlaylistModel', learner_playlist_model),
+            ('PendingDeletionRequestModel', pending_deletion_request_model),
+            ('UserAuthDetailsModel', user_auth_details_model),
+            ('UserBulkEmailsModel', user_bulk_emails_model),
+            ('UserContributionsModel', user_contributions_model),
+            ('UserContributionRightsModel', user_contribution_rights_model),
+            ('UserEmailPreferencesModel', user_email_preferences_model),
+            ('UserRecentChangesBatchModel', user_recent_changes_batch_model),
+            ('UserStatsModel', user_stats_model),
+            ('UserSubscribersModel', user_subscribers_model),
+            ('UserSubscriptionsModel', user_subscriptions_model),
+            ('CollectionProgressModel', collection_progress_model),
+            ('ExpUserLastPlaythroughModel', exp_user_last_playthrough_model),
+            ('StoryProgressModel', story_progress_model),
+            ('UserContributionProficiencyModel', (
+                user_contribution_proficiency_model)),
+            ('UserIdentifiersModel', user_identifiers_model),
+            ('UserSkillMasteryModel', user_skill_mastery_model),
+            ('ExplorationUserDataModel', exploration_user_data_model)
+        ]
+
+        for model_name, model in models_linked_with_user_settings_model:
+            if model is not None:
+                user_dates_list.append((model_name, model.last_updated))
+                user_dates_list.append((model_name, model.created_on))
+
+        # Following models had some additional date/time fields.
+        if user_subscriptions_model is not None:
+            user_dates_list.append(
+                (
+                    'UserSubscriptionsModel',
+                    user_subscriptions_model.last_checked
+                )
+            )
+
+        if exploration_user_data_model is not None:
+            user_dates_list.append(
+                (
+                    'ExplorationUserDataModel',
+                    exploration_user_data_model.rated_on
+                )
+            )
+            user_dates_list.append(
+                (
+                    'ExplorationUserDataModel',
+                    exploration_user_data_model.draft_change_list_last_updated
+                )
+            )
+
+        filtered_user_dates_list = [
+            (model_name, date) for model_name, date in user_dates_list
+            if date is not None
+        ]
+
+        model_name, min_date = min(filtered_user_dates_list, key=lambda x: x[1])
+        if min_date < user_settings_model.created_on:
+            user_settings_model.created_on = min_date
+            user_settings_model.update_timestamps(
+                update_last_updated_time=False)
+            user_settings_model.put()
```

Comment:
For now, how about running this more as an audit job to get info? If you find a discrepancy, output the uid and enough information to debug the discrepancy (e.g. "user xxx created on yyy but field zzz of model www is ttt"). You can even make a list of all the datetimes for all the models if it's useful for analysis. Let's then look at the results before finalizing on what to do (unless this analysis has already been done previously?).

Reply:
Also, if we know there are a lot of these, then you can do something like: generate random number in [0,1], if it's < p (for some chosen value of p) then yield full thing, else just yield an indicator that it was updated.

Reply:
Yes, we know there is a very high number of `UserSettingsModel` that have the `created_on` field broken. I don't think that this kind of output you mention above will help is with this issue anyhow, but I think that just reporting the data (as the yields are set up now) and skipping actually putting the data, might be a good first run.

Reply:
The only thing that we can maybe try to confirm that there are no discrepancies after June 2020 (when the issue was fix by #9252).

Reply:
Would it be useful to have the data available so that we can make a hypothesis and verify it? E.g. we might find, from such an analysis, that the source of truth should be model XXX, and then we can do a check to verify that XXX always has the lowest recorded time among all related models, and then use it as the primary source.

Reply:
Yeah, I see, this is almost what we have implemented, only we do no track the field that was used.

@sajalasati Could you also which field in some model has the lowest value and disable the modification of the data for the first run of the job?

Reply:
@vojtechjelinek Can you confirm that we just want something like `UserSettingsModel_first_contribution_msec` if this model and field were used for updating, or we want to log all updates like @seanlip suggested intially? (I got a bit confused reading through all the comments, though I think the yield statements we have currently should be sufficient for the analysis)

Reply:
@vojtechjelinek Also, I think I should keep a time delta value, such that for a user_id if `x` = UserSettingsModel's lowest timestamp, and `y` = the lowest timestamp we found out among all models, then we update the `created_on` only when `x-y >= time_delta` (when y was smaller). 

This will help us avoid a lot of unnecessary updates, like currently in `user_services.create_new_user` ([here](https://github.com/oppia/oppia/blob/develop/core/domain/user_services.py#L1276-L1278)), first the UserAuthDetails and UserIdentifiers models are created, and UserSettingsModel is created after these two, but the current job will still consider it for an update. (I know it happens because in the tests the default created user from test_utils was getting updated).

@vojtechjelinek I want you to help me with deciding what that `time_delta` should be? Something like 5 mins looks good?

Reply:
> @vojtechjelinek Can you confirm that we just want something like `UserSettingsModel_first_contribution_msec` if this model and field were used for updating, or we want to log all updates like @seanlip suggested intially? (I got a bit confused reading through all the comments, though I think the yield statements we have currently should be sufficient for the analysis)

Yes, something like `UserSettingsModel_first_contribution_msec`, I think this format will be easiest to analyze.

The delta is a nice catch, I think 5 minutes should be fine.

Reply:
I have made the changes now as per [this](https://github.com/oppia/oppia/pull/11437#discussion_r548483548) comment.

Reply:
> The only thing that we can maybe try to confirm that there are no discrepancies after June 2020 (when the issue was fix by #9252).

@vojtechjelinek Sorry I forgot about this - Do we want to yield error statement for these user_models too i.e some sort of error key for models created after June 2020 which don't have their `created_on` properly set? If yes, then I am planning to have something like `('ERROR_FOR_USER_MODEL', 'user_id')`

Reply:
Let's do it from July 1st 2020, but please even if these are broken keep the info about which model has lower value (like with models that are broken before July) so that we do not lose any important info.

Reply:
(Everywhere by cut-off date I mean July 1st, 2020)

@vojtechjelinek Sorry I did not realise this earlier, but it's difficult to identify user models that were "actually" created on/after July 1st 2020, because currently I don't know which attribute I can trust on to be the source of that information.
- Using `UserSettingsModel.created_on` could result in a lot of **false positives** i.e a model that was actually created before July 1st, 2020, but due to earlier flaws has got a timestamp after this cut-off date.
- Rest of the attributes of `UserSettingsModel` can change over time. `first_contribution_msec` remains fixed but it could be None too. And even when it's not None, it has no correlation with creation time.
- Other models were all added over time and some have been added after the cut-off date too, so we can't really trust them for this information.

Considering all of this, I don't see a direct way of identifying them. The **best way according to me** is - If we find that 
  1. `min_date` (among_all_models) <= `user_settings_model.created_on` - `time_delta`, and
  2. `min_date`  >= `cut-off date`

then we **safely assume that the model was created after cut-off date only** (because even the best `min_date` we are updating it to also lies after `cut-off date`). I think either we use this or drop the check only, as I see no other way around.

Thoughts?

Reply:
@sajalasati Ah, good analysis, I agree with your approach, thanks for the great explanation.